tests/unit/test_llm_utils.py

```diff
@@ -0,0 +1,178 @@
+        df = pd.DataFrame({'text': user_inputs})
+        with self.assertRaises(Exception):
+            get_completed_prompts(base_template, df)
+
+    def test_ft_chat_format_validation(self):
+        valid_chats = [
+            # u/a pattern
+            [
+                {"role": "user", "content": "hi"},
+                {"role": "assistant", "content": "hello"},
+                {"role": "user", "content": "how are you?"},
+                {"role": "assistant", "content": "I'm good, thanks"},
+            ],
+
+            # u/a pattern
+            [
+                {"role": "user", "content": "hi"},
+                {"role": "assistant", "content": "hello"},
+                {"role": "user", "content": "how are you?"},
+            ],
+
+            # s/u/a pattern
+            [
+                {"role": "system", "content": "you are a useful assistant."},
+                {"role": "user", "content": "hello"},
+                {"role": "assistant", "content": "how are you?"},
+            ],
+
+            # s/u/a pattern
+            [
+                {"role": "system", "content": "you are a useful assistant."},
+                {"role": "user", "content": "hello"},
+                {"role": "assistant", "content": "how are you?"},
+                {"role": "user", "content": "I'm good, thanks"},
+            ],
+        ]
+
+        invalid_chats = [
+            # invalid - repeated user
+            [
+                {"role": "user", "content": "hi"},
+                {"role": "user", "content": "hello"},  # this is invalid
+                {"role": "assistant", "content": "how are you?"},
+                {"role": "user", "content": "I'm good, thanks"},
+            ],
+
+            # invalid - repeated assistant
+            [
+                {"role": "user", "content": "hi"},
+                {"role": "assistant", "content": "hello"},
+                {"role": "assistant", "content": "how are you?"},  # this is invalid
+                {"role": "user", "content": "I'm good, thanks"},
+            ],
+
+            # invalid - incorrect system prompt order
+            [
+                {"role": "user", "content": "hi"},
+                {"role": "assistant", "content": "hello"},
+                {"role": "system", "content": "you are a useful assistant."},  # this is invalid
+                {"role": "user", "content": "I'm good, thanks"},
+            ],
+
+            # invalid roles
+            [
+                {"role": "user", "content": "hi"},
+                {"role": "invalid", "content": "this is an invalid role"},
+            ],
+
+            # invalid content
+            [
+                {"role": "user", "content": "hi"},
+                {"role": "assistant", "content": None},  # should always be a string
+            ],
+
+            # invalid - no assistant in the chat
+            [
+                {"role": "user", "content": "hi"},
+            ],
+        ]
+
```

Comment:
minor nit: perhaps we can make invalid_chats and valid_chats a fixture. Stylistic, doesn't block merge